misk-service/src/main/kotlin/misk/CoordinatedService.kt

```diff
@@ -134,15 +139,11 @@ internal class CoordinatedService(
   }
 
   private fun startIfReady() {
-    synchronized(this) {
-      if (state() != State.STARTING) return
+    val canStartInner = state() == State.STARTING && upstreamServices.all { it.isRunning() }
 
-      // If any upstream service or its enhancements are not running, don't start.
-      if (upstreamServices.any { !it.isRunning() }) return
-
-      if (service.state() != State.NEW) return
-
-      // Actually start.
+    // startAsync must be called exactly once
+    val started = innerServiceStarted.getAndSet(canStartInner)
```

Comment:
I don't think this is right. if `canStartInner` is false, it may overwrite the `started` flag to false.
e.g.
`startIfReady` is called 3 times in different threads:
* Thread 1 gets to this point with `canStartInner = true`, then sets the `innerServiceStarted` flag and starts the service.
* Thread 2 gets to this point with `canStartInner = false` due to race, and sets the `innerServiceStarted = false`
* Thread 3 gets to this point with `canStartInner = true` and starts the service again, violating the contract

I think this should be combined into `if (canStartInner && !innerServiceStarted.getAndSet(true))`

Reply:
The assumption is `canStartInner` will never go from `true` to `false`while in the `STARTING` state - but maybe I should assert this in some way or add a guard for this case.